File: src/main/kotlin/oop/Address.java
package oop;

import java.util.Objects;

public class Address {
    private String city;
    private String street;
    private int houseNum;

    public Address(String city, String street, int houseNum) {
        this.city = city;
        this.street = street;
        this.houseNum = houseNum;
    }

    public String getCity() {
        return city;
    }

    public void setCity(String city) {
        this.city = city;
    }

    public String getStreet() {
        return street;
    }

    public void setStreet(String street) {
        this.street = street;
    }

    public int getHouseNum() {
        return houseNum;
    }

    public void setHouseNum(int houseNum) {
        this.houseNum = houseNum;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Address address = (Address) o;
        return houseNum == address.houseNum && Objects.equals(city, address.city) && Objects.equals(street, address.street);
    }

    @Override
    public int hashCode() {
        return Objects.hash(city, street, houseNum);
    }

    public static void main(String[] args) {
        Address address1 = new Address("Moscow", "Otradnaya", 18);
        Address address2 = new Address("Moscow", "Otradnaya", 18);

        if (address1.equals(address2)) {
            System.out.println("Equal!");
        } else {
            System.out.println("Not equal!");
        }
    }
}

File: src/main/kotlin/oop/dataclass/equalsAndHashcode/Address.kt
package oop.dataclass.equalsAndHashcode

class Address (val city: String, val street: String, val houseNum: Int) {

    fun copy(city: String = this.city, street: String = this.street, houseNum: Int = this.houseNum): Address {
        return Address(city, street, houseNum)
    }

    operator fun component1() = city
    operator fun component2() = street
    operator fun component3() = houseNum

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Address

        if (city != other.city) return false
        if (street != other.street) return false
        if (houseNum != other.houseNum) return false

        return true
    }

    override fun hashCode(): Int {
        var result = city.hashCode()
        result = 31 * result + street.hashCode()
        result = 31 * result + houseNum
        return result
    }

}

fun main() {
    val address1 = Address("Moscow", "Zelenaya", 12)
    val address2 = address1.copy()

    println(address1)
    println(address2)

    println(address1.hashCode())
    println(address2.hashCode())

    println(address1 == address2)

    val (city, street, houseNum) = address1
}